### ISSUE TITLE
能提供下论文吗，付费看不了，想学习下？

### ISSUE THREAD

**Issue by xingguang12** [ZHjiuang/yolov5-ghostv2-SIOU-CA-WBF-V6.2#1]

**Comment by ZHjiuang** [comment]
[An improved target detection method based on YOLOv5 in natural_compressed.pdf](https://github.com/ZHjiuang/yolov5-ghostnetv2-SIOU-CA-WBF-V6.2/files/14469630/An.improved.target.detection.method.based.on.YOLOv5.in.natural_compressed.pdf)

**Comment by xingguang12** [comment]
> [An improved target detection method based on YOLOv5 in natural_compressed.pdf](https://github.com/ZHjiuang/yolov5-ghostnetv2-SIOU-CA-WBF-V6.2/files/14469630/An.improved.target.detection.method.based.on.YOLOv5.in.natural_compressed.pdf)

好的特别感谢，怎么感觉代码和原始的ghostnetv2不一样啊，原始用的是GhostModuleV2，您这边用的是GhostConv
![20240302224714](https://github.com/ZHjiuang/yolov5-ghostnetv2-SIOU-CA-WBF-V6.2/assets/57353111/8fd9eb12-df55-4f06-945f-e1e5f0bea79c)
class GhostModuleV2(nn.Module):
    def __init__(self, inp, oup, kernel_size=1, ratio=2, dw_size=3, stride=1, relu=True,mode=None,args=None):
        super(GhostModuleV2, self).__init__()
        self.mode=mode
        self.gate_fn=nn.Sigmoid()

        if self.mode in ['original']:
            self.oup = oup
            init_channels = math.ceil(oup / ratio) 
            new_channels = init_channels*(ratio-1)
            self.primary_conv = nn.Sequential(  
                nn.Conv2d(inp, init_channels, kernel_size, stride, kernel_size//2, bias=False),
                nn.BatchNorm2d(init_channels),
                nn.ReLU(inplace=True) if relu else nn.Sequential(),
            )
            self.cheap_operation = nn.Sequential(
                nn.Conv2d(init_channels, new_channels, dw_size, 1, dw_size//2, groups=init_channels, bias=False),
                nn.BatchNorm2d(new_channels),
                nn.ReLU(inplace=True) if relu else nn.Sequential(),
            )
        elif self.mode in ['attn']: 
            self.oup = oup
            init_channels = math.ceil(oup / ratio) 
            new_channels = init_channels*(ratio-1)
            self.primary_conv = nn.Sequential(  
                nn.Conv2d(inp, init_channels, kernel_size, stride, kernel_size//2, bias=False),
                nn.BatchNorm2d(init_channels),
                nn.ReLU(inplace=True) if relu else nn.Sequential(),
            )
            self.cheap_operation = nn.Sequential(
                nn.Conv2d(init_channels, new_channels, dw_size, 1, dw_size//2, groups=init_channels, bias=False),
                nn.BatchNorm2d(new_channels),
                nn.ReLU(inplace=True) if relu else nn.Sequential(),
            ) 
            self.short_conv = nn.Sequential( 
                nn.Conv2d(inp, oup, kernel_size, stride, kernel_size//2, bias=False),
                nn.BatchNorm2d(oup),
                nn.Conv2d(oup, oup, kernel_size=(1,5), stride=1, padding=(0,2), groups=oup,bias=False),
                nn.BatchNorm2d(oup),
                nn.Conv2d(oup, oup, kernel_size=(5,1), stride=1, padding=(2,0), groups=oup,bias=False),
                nn.BatchNorm2d(oup),
            ) 
      
    def forward(self, x):
        if self.mode in ['original']:
            x1 = self.primary_conv(x)
            x2 = self.cheap_operation(x1)
            out = torch.cat([x1,x2], dim=1)
            return out[:,:self.oup,:,:]         
        elif self.mode in ['attn']:  
            res=self.short_conv(F.avg_pool2d(x,kernel_size=2,stride=2))  
            x1 = self.primary_conv(x)
            x2 = self.cheap_operation(x1)
            out = torch.cat([x1,x2], dim=1)
            return out[:,:self.oup,:,:]*F.interpolate(self.gate_fn(res),size=(out.shape[-2],out.shape[-1]),mode='nearest') 
我看不太懂这个ghostnetv2的代码

**Comment by ZHjiuang** [comment]
> > [An improved target detection method based on YOLOv5 in natural_compressed.pdf](https://github.com/ZHjiuang/yolov5-ghostnetv2-SIOU-CA-WBF-V6.2/files/14469630/An.improved.target.detection.method.based.on.YOLOv5.in.natural_compressed.pdf)
> 
> 好的特别感谢，怎么感觉代码和原始的ghostnetv2不一样啊，原始用的是GhostModuleV2，您这边用的是GhostConv ![20240302224714](https://private-user-images.githubusercontent.com/57353111/309480372-8fd9eb12-df55-4f06-945f-e1e5f0bea79c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk0MzE5OTAsIm5iZiI6MTcwOTQzMTY5MCwicGF0aCI6Ii81NzM1MzExMS8zMDk0ODAzNzItOGZkOWViMTItZGY1NS00ZjA2LTk0NWYtZTFlNWYwYmVhNzljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAzVDAyMDgxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2Yjk2NDFlOTM5NzFjNjk4OGJjZjJmMzliOGVlMDY5ODNhMmU2NDkzNTRlYzhmZTQzODUzMWIyNjQ4NmY2ZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YDbhguxVdJdBmwY_wQ2QsGHuLQK-1i5yRblzDivKV8k) class GhostModuleV2(nn.Module): def **init**(self, inp, oup, kernel_size=1, ratio=2, dw_size=3, stride=1, relu=True,mode=None,args=None): super(GhostModuleV2, self).**init**() self.mode=mode self.gate_fn=nn.Sigmoid()
> 
> ```
>     if self.mode in ['original']:
>         self.oup = oup
>         init_channels = math.ceil(oup / ratio) 
>         new_channels = init_channels*(ratio-1)
>         self.primary_conv = nn.Sequential(  
>             nn.Conv2d(inp, init_channels, kernel_size, stride, kernel_size//2, bias=False),
>             nn.BatchNorm2d(init_channels),
>             nn.ReLU(inplace=True) if relu else nn.Sequential(),
>         )
>         self.cheap_operation = nn.Sequential(
>             nn.Conv2d(init_channels, new_channels, dw_size, 1, dw_size//2, groups=init_channels, bias=False),
>             nn.BatchNorm2d(new_channels),
>             nn.ReLU(inplace=True) if relu else nn.Sequential(),
>         )
>     elif self.mode in ['attn']: 
>         self.oup = oup
>         init_channels = math.ceil(oup / ratio) 
>         new_channels = init_channels*(ratio-1)
>         self.primary_conv = nn.Sequential(  
>             nn.Conv2d(inp, init_channels, kernel_size, stride, kernel_size//2, bias=False),
>             nn.BatchNorm2d(init_channels),
>             nn.ReLU(inplace=True) if relu else nn.Sequential(),
>         )
>         self.cheap_operation = nn.Sequential(
>             nn.Conv2d(init_channels, new_channels, dw_size, 1, dw_size//2, groups=init_channels, bias=False),
>             nn.BatchNorm2d(new_channels),
>             nn.ReLU(inplace=True) if relu else nn.Sequential(),
>         ) 
>         self.short_conv = nn.Sequential( 
>             nn.Conv2d(inp, oup, kernel_size, stride, kernel_size//2, bias=False),
>             nn.BatchNorm2d(oup),
>             nn.Conv2d(oup, oup, kernel_size=(1,5), stride=1, padding=(0,2), groups=oup,bias=False),
>             nn.BatchNorm2d(oup),
>             nn.Conv2d(oup, oup, kernel_size=(5,1), stride=1, padding=(2,0), groups=oup,bias=False),
>             nn.BatchNorm2d(oup),
>         ) 
>   
> def forward(self, x):
>     if self.mode in ['original']:
>         x1 = self.primary_conv(x)
>         x2 = self.cheap_operation(x1)
>         out = torch.cat([x1,x2], dim=1)
>         return out[:,:self.oup,:,:]         
>     elif self.mode in ['attn']:  
>         res=self.short_conv(F.avg_pool2d(x,kernel_size=2,stride=2))  
>         x1 = self.primary_conv(x)
>         x2 = self.cheap_operation(x1)
>         out = torch.cat([x1,x2], dim=1)
>         return out[:,:self.oup,:,:]*F.interpolate(self.gate_fn(res),size=(out.shape[-2],out.shape[-1]),mode='nearest') 
> ```
> 
> 我看不太懂这个ghostnetv2的代码

![2024-03-03 100800](https://github.com/ZHjiuang/yolov5-ghostnetv2-SIOU-CA-WBF-V6.2/assets/91889803/137f17cc-f0d0-4b13-9817-5a8b4b76f296)
将这两个部分写到DFC类里面了，并且只用了`self.mode = ['attn']`，原代码涉及到的`primary_conv`与`cheap_operation`部分用GhostConv替代了，其他的结构与原来是类似的。
